ember-server/src/main/scala/org/http4s/ember/server/internal/WebSocketHelpers.scala

```diff
@@ -0,0 +1,278 @@
+                frame <- F.fromEither(WebSocketFrame.Close(frame.closeCode))
+                _ <- writeFrame(frame)
+                _ <- closeState.set(BothClosed)
+              } yield None
+            case _ => F.pure(None)
+          }
+        case x => F.pure(Some(x))
+      }
+  }
+
+  private def encodeFrames[F[_]](frameTranscoder: FrameTranscoder): Pipe[F, WebSocketFrame, Byte] =
+    stream =>
+      stream.flatMap { frame =>
+        val chunks = frameTranscoder.frameToBuffer(frame).map { buffer =>
+          // TODO followup: improve the buffering here
+          val bytes = new Array[Byte](buffer.remaining())
+          buffer.get(bytes)
+          Chunk.array(bytes)
+        }
+        Stream
+          .iterable(chunks)
+          .flatMap(Stream.chunk(_))
+      }
+
+  private def decodeFrames[F[_]](frameTranscoder: FrameTranscoder)(implicit
+      F: Async[F]): Pipe[F, Byte, WebSocketFrame] = stream => {
+    def go(rest: Stream[F, Byte], acc: Array[Byte]): Pull[F, WebSocketFrame, Unit] =
+      rest.pull.uncons.flatMap {
+        case Some((chunk, next)) =>
+          val buffer = acc ++ chunk.toArray[Byte]
+          val byteBuffer = ByteBuffer.wrap(buffer)
+          Pull
+            .eval(F.delay(frameTranscoder.bufferToFrame(byteBuffer)))
+            .flatMap { value =>
+              // TODO followup: improve this buffering
+              if (value != null) {
+                val remaining = new Array[Byte](byteBuffer.remaining())
+                byteBuffer.get(remaining)
+                Pull.output1(value) >> go(next, remaining)
+              } else {
+                go(next, buffer)
+              }
+            }
+        case None =>
+          // TODO followup: sometimes the peer closes connection before stream can interrupt itself
+          Pull.raiseError(EndOfStreamError())
+      }
+
+    go(stream, Array.emptyByteArray).void.stream
+  }
+
+  private def clientHandshake[F[_]](req: Request[F]): Either[ClientHandshakeError, String] = {
+    val connection = req.headers.get[Connection] match {
+      case Some(header) if header.hasUpgrade => Either.unit
+      case _ => Left(UpgradeRequired)
+    }
+
+    val upgrade = req.headers.get[Upgrade] match {
+      case Some(header) if header.values.contains_(webSocketProtocol) => Either.unit
+      case _ => Left(UpgradeRequired)
+    }
+
+    val version = req.headers.get[`Sec-WebSocket-Version`] match {
+      case Some(header) if header.version == supportedWebSocketVersion => Either.unit
+      case Some(header) => Left(UnsupportedVersion(supportedWebSocketVersion, header.version))
+      case None => Left(VersionNotFound)
+    }
+
+    val key = req.headers.get[`Sec-WebSocket-Key`] match {
+      case Some(header) => Right(header.value)
+      case None => Left(KeyNotFound)
+    }
+
+    (connection, upgrade, version, key).mapN { case (_, _, _, key) => key }
+  }
+
+  private def serverHandshake[F[_]](value: String)(implicit F: MonadThrow[F]): F[ByteVector] =
```

Comment:
This was `Sync` in 0.22, but all we're doing is suspending errors into `F`, I think.  I was hoping it got us out of the `Async` constraint mentioned elsewhere in this PR, but nope.